### PR TITLE
coverage: Specify modules to run coverage for

### PR DESCRIPTION
**Pull request by stephenfin**
### Feature or Bugfix

Feature

### Purpose

Currently, the coverage builder lets you check for partially documented modules, but there is no mechanism to identify totally undocumented modules. Resolve this by introducing a new `coverage_modules` config option. This is a list of modules that should be documented somewhere within the documentation tree. Any modules that are specified in the configuration value but are not documented anywhere will result in a warning. Likewise, any modules that are not in the config option but are documented somewhere will result in a warning.

> **Note**
> I have a WIP series locally that will complete the documentation for this module and close out #1288, but I'm keeping that out of this to avoid making it any more complicated than necessary.

### Detail

(none)

### Relates

(none)

### CODE REVIEW

**Comment by stephenfin** [comment]
I'll wait for the test to re-run, but it looks like the CI failures are flakes and this is otherwise passing as expected.

@AA-Turner Who's the best person other than me to review this and try get it into 7.3/8.0?

**Comment by stephenfin** [comment]
fwiw, I came up with this after spotting the gap while trying to use coverage for a project I contribute to. [After testing](https://review.opendev.org/c/openstack/openstacksdk/+/891690), it seemed to do what I expected and highlighted a whole load of missing documentation.

File: doc/usage/extensions/coverage.rst
Comment (picnixz) [inline doc/usage/extensions/coverage.rst:9]: Didn't you just write it :) ?

File: sphinx/ext/coverage.py
Comment (picnixz) [inline sphinx/ext/coverage.py]: I think we can skip them by using a regex instead so let's pick them by default.

File: sphinx/ext/coverage.py
Comment (picnixz) [inline sphinx/ext/coverage.py]: Don't you want to exclude a submodule if the latter matches a regexpr?

File: sphinx/ext/coverage.py
Comment (picnixz) [inline sphinx/ext/coverage.py]: I think module_names would be better.

File: sphinx/ext/coverage.py
Comment (picnixz) [inline sphinx/ext/coverage.py]: Since people like to suppress warnings, could you also decide for a type/subtype for the log record and add it to the existing ones.

**Review by picnixz — COMMENTED** [review]
Do we need some extra tests?

File: sphinx/ext/coverage.py
Comment (stephenfin) [reply sphinx/ext/coverage.py]: wfm

File: sphinx/ext/coverage.py
Comment (stephenfin) [reply sphinx/ext/coverage.py]: Good point. Done.

File: sphinx/ext/coverage.py
Comment (stephenfin) [reply sphinx/ext/coverage.py]: Agreed. Done.

File: sphinx/ext/coverage.py
Comment (stephenfin) [reply sphinx/ext/coverage.py]: My one disagreement: none of the other `logger.warning` calls in here do this, so I think we want to do it for all or none. Can I do this as a follow-up? I've added a TODO.

Comment (picnixz) [reply sphinx/ext/coverage.py]: Sure! Overall, we need to add those type/subtype more or less everywhere where warnings could be annoying because people request the ability to suppress them (and to turn them as non-errors), so you could have a separate issue for warnings regarding coverage-related things.

**Comment by stephenfin** [comment]
> Do we need some extra tests?

Yes, we do. Added now

File: sphinx/ext/coverage.py
Comment (picnixz) [inline sphinx/ext/coverage.py]: Actually, do you want the function to be part of the public API? or could we rename it `_load_modules`?

Comment (stephenfin) [reply sphinx/ext/coverage.py]: Fair point. Done.

File: sphinx/ext/coverage.py
Comment (picnixz) [inline sphinx/ext/coverage.py]:
```suggestion
        if all_objects:
```

File: sphinx/ext/coverage.py
Comment (picnixz) [inline sphinx/ext/coverage.py]:
```suggestion
        all_modules = frozenset(self.py_documented.keys() | self.py_undocumented.keys())
```

File: tests/roots/test-ext-coverage/grog/coverage_missing.py
Comment (picnixz) [inline tests/roots/test-ext-coverage/grog/coverage_missing.py]:
```suggestion
```

File: tests/test_ext_coverage.py
Comment (picnixz) [inline tests/test_ext_coverage.py]: Any reason for this change?

**Review by picnixz — COMMENTED** [review]
Just so that you know, I didn't review the doc content.

File: sphinx/ext/coverage.py
Comment (stephenfin) [reply sphinx/ext/coverage.py]: I was pretty sure this was done for a reason. Something like this is what had been done elsewhere. I can't remember now though so...

File: tests/test_ext_coverage.py
Comment (stephenfin) [reply tests/test_ext_coverage.py]: I think this is an artefact from an earlier revision of the patch. It's unnecessary but also harmless. I'll take it out if I need to respin.

**Comment by stephenfin** [comment]
> Just so that you know, I didn't review the doc content.

Thanks. I think I have addressed everything now?

**Comment by picnixz** [comment]
I think so. At least I don't have anything more in mind, except the todo for the doc but as you mentioned, it'll be another PR

File: sphinx/ext/coverage.py
Comment (AA-Turner) [inline sphinx/ext/coverage.py]: I think this warning only captures half of what you noted -- modules that are in ``coverage_modules`` but not in ``seen_modules`` don't seem to get a warning.

```suggestion
    # if there are additional modules then we warn (but still scan)
    if additional_modules := set(seen_modules) - modules:
```

File: sphinx/ext/coverage.py
Comment (AA-Turner) [inline sphinx/ext/coverage.py]: I wonder if we'd be able to use ``importlib`` (e.g. [``importlib.util.find_spec``](https://docs.python.org/dev/library/importlib.html#importlib.util.find_spec); [``importlib.util.spec_from_file_location``](https://docs.python.org/dev/library/importlib.html#importlib.util.spec_from_file_location)) to avoid importing the modules, given we only want to find their names and subpackages.

**Review by AA-Turner — APPROVED** [review]
Looks good, two notes

A

File: sphinx/ext/coverage.py
Comment (stephenfin) [reply sphinx/ext/coverage.py]: Does it matter? We won't import modules that we don't want to document as we check the module name first, and we import all the other modules anyway via later calls to `import_module`.

File: sphinx/ext/coverage.py
Comment (stephenfin) [reply sphinx/ext/coverage.py]: They weren't getting a warning but they were getting captured in the report. I've added the warning nonetheless.

**Comment by stephenfin** [comment]
@AA-Turner this is done now

File: doc/usage/extensions/coverage.rst
Comment (AA-Turner) [inline doc/usage/extensions/coverage.rst:9]: Will this TODO remain?

File: doc/usage/extensions/coverage.rst
Comment (AA-Turner) [inline doc/usage/extensions/coverage.rst]:
```suggestion
   .. versionadded:: 7.2
```

File: doc/usage/extensions/coverage.rst
Comment (stephenfin) [reply doc/usage/extensions/coverage.rst:9]: Not long-term. I have a separate series but GitHub doesn't let me stack PRs. I'll push once this is merged.

**Comment by AA-Turner** [comment]
Thanks!

A

**Pull request merged**
